File: src/minima_harness/session/store.py
from __future__ import annotations

import logging
import os
from dataclasses import dataclass
from pathlib import Path

from minima_harness.session.format import EntryType, SessionEntry, new_id

_log = logging.getLogger("minima_harness.session")


class SessionStore:
    """Append-only JSONL session tree. File-backed or in-memory (``--no-session``)."""

    def __init__(self, path: Path | None, *, display_name: str | None = None) -> None:
        self._path = path
        self._mem: list[SessionEntry] = []
        self._tip: str | None = None
        self.display_name = display_name
        if path is not None and path.exists():
            self._reload()

    @classmethod
    def file_backed(cls, path: Path, *, display_name: str | None = None) -> SessionStore:
        return cls(path, display_name=display_name)

    @classmethod
    def in_memory(cls) -> SessionStore:
        return cls(None)

    @property
    def path(self) -> Path | None:
        return self._path

    @property
    def persistent(self) -> bool:
        return self._path is not None

    @property
    def entries(self) -> list[SessionEntry]:
        return list(self._mem)

    @property
    def tip(self) -> str | None:
        return self._tip

    def append(
        self, entry_type: EntryType, payload: dict, *, label: str | None = None
    ) -> SessionEntry:
        entry = SessionEntry(
            id=new_id(),
            parent_id=self._tip,
            type=entry_type,
            payload=payload,
            label=label,
        )
        self._mem.append(entry)
        self._tip = entry.id
        if self._path is not None:
            try:
                with self._path.open("a", encoding="utf-8") as fh:
                    fh.write(entry.model_dump_json() + "\n")
            except OSError:  # noqa: BLE001 - disk failure must not kill the turn
                _log.warning("session_append_failed", exc_info=True)
        return entry

    def set_tip(self, entry_id: str) -> None:
        """Branch: continue the next append from ``entry_id`` (must already exist)."""
        if not any(e.id == entry_id for e in self._mem):
            raise KeyError(f"unknown entry id: {entry_id}")
        self._tip = entry_id

    def path_to(self, entry_id: str) -> list[SessionEntry]:
        """Root → entry_id path (inclusive). Raises KeyError if unknown."""
        by_id = {e.id: e for e in self._mem}
        if entry_id not in by_id:
            raise KeyError(f"unknown entry id: {entry_id}")
        out: list[SessionEntry] = []
        cur: str | None = entry_id
        while cur is not None and cur in by_id:
            out.append(by_id[cur])
            cur = by_id[cur].parent_id
        out.reverse()
        return out

    def children_map(self) -> dict[str | None, list[str]]:
        """parentId → child ids in insertion order (root key is None)."""
        cm: dict[str | None, list[str]] = {}
        for e in self._mem:
            cm.setdefault(e.parent_id, []).append(e.id)
        return cm

    def _write_path(self, dest: Path, entries: list[SessionEntry]) -> None:
        dest.parent.mkdir(parents=True, exist_ok=True)
        with dest.open("w", encoding="utf-8") as fh:
            for e in entries:
                fh.write(e.model_dump_json() + "\n")

    def fork_to(self, dest: Path, *, from_entry_id: str) -> SessionStore:
        """Copy the root→from_entry_id path into a new session file."""
        path = self.path_to(from_entry_id)
        self._write_path(dest, path)
        return SessionStore.file_backed(dest)

    def clone_to(self, dest: Path) -> SessionStore:
        """Copy the current branch (root→tip) into a new session file."""
        if self._tip is None:
            self._write_path(dest, [])
            return SessionStore.file_backed(dest)
        return self.fork_to(dest, from_entry_id=self._tip)

    def _reload(self) -> None:
        assert self._path is not None
        self._mem = []
        for line in self._path.read_text(encoding="utf-8").splitlines():
            if not line.strip():
                continue
            try:
                self._mem.append(SessionEntry.model_validate_json(line))
            except Exception:  # noqa: BLE001 - one bad line must not lose the session
                _log.warning("session_skipped_malformed_line")
        self._tip = self._mem[-1].id if self._mem else None


@dataclass(slots=True)
class SessionSummary:
    session_id: str
    path: Path
    display_name: str | None
    mtime: float
    n_entries: int


class SessionManager:
    """Discovers/creates session files under ``<sessions_dir>/<cwd-slug>/<uuid>.jsonl``."""

    def __init__(self, sessions_dir: Path | None = None) -> None:
        base = sessions_dir or Path.home() / ".minima-harness" / "sessions"
        self._base = Path(base)

    def slug_for(self, directory: Path) -> str:
        directory = Path(directory).resolve()
        slug = str(directory).replace(os.sep, "-").replace("/", "-")
        return slug.lstrip("-") or "root"

    def _dir_for(self, directory: Path) -> Path:
        d = self._base / self.slug_for(directory)
        d.mkdir(parents=True, exist_ok=True)
        return d

    def new(self, directory: Path, *, name: str | None = None) -> SessionStore:
        sid = new_id()
        path = self._dir_for(directory) / f"{sid}.jsonl"
        return SessionStore.file_backed(path, display_name=name)

    def open(
        self,
        directory: Path,
        *,
        session_id: str | None = None,
        no_session: bool = False,
    ) -> SessionStore:
        if no_session:
            return SessionStore.in_memory()
        if session_id:
            for s in self.list_sessions(directory):
                if s.session_id.startswith(session_id) or session_id.startswith(s.session_id):
                    return SessionStore.file_backed(s.path, display_name=s.display_name)
            raise FileNotFoundError(f"no session matching id: {session_id}")
        recent = self.most_recent(directory)
        if recent is not None:
            return SessionStore.file_backed(recent.path, display_name=recent.display_name)
        return self.new(directory)

    def most_recent(self, directory: Path) -> SessionSummary | None:
        sessions = self.list_sessions(directory)
        return max(sessions, key=lambda s: s.mtime) if sessions else None

    def list_sessions(self, directory: Path) -> list[SessionSummary]:
        d = self._base / self.slug_for(directory)
        if not d.exists():
            return []
        out: list[SessionSummary] = []
        for p in sorted(d.glob("*.jsonl")):
            try:
                lines = p.read_text(encoding="utf-8").splitlines()
                count = sum(1 for ln in lines if ln.strip())
            except OSError:  # noqa: BLE001
                continue
            out.append(
                SessionSummary(
                    session_id=p.stem,
                    path=p,
                    display_name=None,
                    mtime=p.stat().st_mtime,
                    n_entries=count,
                )
            )
        return out

File: packages/tui/src/session/store.ts
/**
 * Append-only JSONL session tree + SessionManager.
 *
 * Port of the Python harness's session/store.py + format.py. One node per turn, linked by
 * parentId, persisted as one JSON line per entry. File-backed or in-memory (--no-session).
 * A disk failure on append is logged-and-swallowed so it never kills a turn.
 */

import { randomUUID } from "node:crypto";
import { appendFile, mkdir, readFile, readdir, stat, writeFile } from "node:fs/promises";
import { homedir } from "node:os";
import { basename, dirname, join, resolve as resolvePath } from "node:path";

export type EntryType = "user" | "assistant" | "tool" | "system" | "goal";

export interface SessionEntry {
  id: string;
  parentId: string | null;
  type: EntryType;
  ts: number;
  payload: Record<string, unknown>;
  label?: string | null;
}

/** A short, unique entry/session id (first 12 hex chars of uuid4). */
export function newId(): string {
  return randomUUID().replace(/-/g, "").slice(0, 12);
}

export function nowTs(): number {
  return Date.now() / 1000;
}

/** Compact relative age: just now / 5m ago / 2h ago / 3d ago / 5w ago. */
export function formatAge(ts: number, now: number = nowTs()): string {
  if (!ts || ts <= 0) return "?";
  const delta = Math.max(0, now - ts);
  if (delta < 60) return "just now";
  if (delta < 3600) return `${Math.floor(delta / 60)}m ago`;
  if (delta < 86400) return `${Math.floor(delta / 3600)}h ago`;
  if (delta < 86400 * 7) return `${Math.floor(delta / 86400)}d ago`;
  return `${Math.floor(delta / (86400 * 7))}w ago`;
}

export class SessionStore {
  private readonly path: string | null;
  private mem: SessionEntry[] = [];
  private tipId: string | null = null;
  displayName: string | null;
  readonly persistent: boolean;

  private constructor(path: string | null, displayName?: string | null) {
    this.path = path;
    this.persistent = path !== null;
    this.displayName = displayName ?? null;
  }

  static async fileBacked(path: string, displayName?: string | null): Promise<SessionStore> {
    const s = new SessionStore(path, displayName);
    await s.reload();
    return s;
  }

  static inMemory(): SessionStore {
    return new SessionStore(null);
  }

  get entries(): SessionEntry[] {
    return [...this.mem];
  }

  get tip(): string | null {
    return this.tipId;
  }

  /** Append one entry; returns it. Persists to disk best-effort (never throws). */
  async append(
    type: EntryType,
    payload: Record<string, unknown>,
    label?: string,
  ): Promise<SessionEntry> {
    const entry: SessionEntry = {
      id: newId(),
      parentId: this.tipId,
      type,
      ts: nowTs(),
      payload,
      label: label ?? null,
    };
    this.mem.push(entry);
    this.tipId = entry.id;
    if (this.path) {
      try {
        await appendFile(this.path, `${JSON.stringify(entry)}\n`, "utf8");
      } catch {
        // disk failure must not kill the turn
      }
    }
    return entry;
  }

  /** Branch: continue the next append from `entryId` (must already exist). */
  setTip(entryId: string): void {
    if (!this.mem.some((e) => e.id === entryId)) throw new Error(`unknown entry id: ${entryId}`);
    this.tipId = entryId;
  }

  /** Root → entryId path (inclusive). Throws if unknown. */
  pathTo(entryId: string): SessionEntry[] {
    const byId = new Map(this.mem.map((e) => [e.id, e]));
    const target = byId.get(entryId);
    if (!target) throw new Error(`unknown entry id: ${entryId}`);
    const out: SessionEntry[] = [];
    let cur: string | null = entryId;
    while (cur !== null) {
      const e = byId.get(cur);
      if (!e) break;
      out.push(e);
      cur = e.parentId;
    }
    return out.reverse();
  }

  /** parentId → child ids in insertion order (root key is null). */
  childrenMap(): Map<string | null, string[]> {
    const cm = new Map<string | null, string[]>();
    for (const e of this.mem) {
      const list = cm.get(e.parentId) ?? [];
      list.push(e.id);
      cm.set(e.parentId, list);
    }
    return cm;
  }

  /** Copy the root→fromEntryId path into a new session file. */
  async forkTo(dest: string, fromEntryId: string): Promise<SessionStore> {
    const path = this.pathTo(fromEntryId);
    await writeEntries(dest, path);
    return SessionStore.fileBacked(dest);
  }

  /** Copy the current branch (root→tip) into a new session file. */
  async cloneTo(dest: string): Promise<SessionStore> {
    if (this.tipId === null) {
      await writeEntries(dest, []);
      return SessionStore.fileBacked(dest);
    }
    return this.forkTo(dest, this.tipId);
  }

  private async reload(): Promise<void> {
    if (!this.path) return;
    try {
      const text = await readFile(this.path, "utf8");
      for (const line of text.split(/\r?\n/)) {
        if (!line.trim()) continue;
        try {
          this.mem.push(JSON.parse(line) as SessionEntry);
        } catch {
          // one bad line must not lose the session
        }
      }
      this.tipId = this.mem.length ? this.mem[this.mem.length - 1]!.id : null;
    } catch {
      // missing file
    }
  }
}

async function writeEntries(dest: string, entries: SessionEntry[]): Promise<void> {
  await mkdir(dirname(dest), { recursive: true });
  const body = entries.map((e) => JSON.stringify(e)).join("\n");
  await writeFile(dest, entries.length ? `${body}\n` : "", "utf8");
}

export interface SessionSummary {
  sessionId: string;
  path: string;
  displayName: string | null;
  mtime: number;
  nEntries: number;
  created: number;
}

/** Discovers/creates session files under `<base>/<cwd-slug>/<uuid>.jsonl`. */
export class SessionManager {
  private readonly base: string;
  constructor(base?: string) {
    this.base = base ?? join(homedir(), ".minima-harness", "sessions");
  }

  slugFor(directory: string): string {
    const resolved = resolvePath(directory);
    return resolved.replace(/[/\\]/g, "-").replace(/^-+/, "") || "root";
  }

  private async dirFor(directory: string): Promise<string> {
    const d = join(this.base, this.slugFor(directory));
    await mkdir(d, { recursive: true });
    return d;
  }

  async new(directory: string, name?: string): Promise<SessionStore> {
    const sid = newId();
    const path = join(await this.dirFor(directory), `${sid}.jsonl`);
    return SessionStore.fileBacked(path, name ?? null);
  }

  async open(
    directory: string,
    opts: { sessionId?: string; noSession?: boolean } = {},
  ): Promise<SessionStore> {
    if (opts.noSession) return SessionStore.inMemory();
    if (opts.sessionId) {
      const sessions = await this.listSessions(directory);
      for (const s of sessions) {
        if (s.sessionId.startsWith(opts.sessionId) || opts.sessionId.startsWith(s.sessionId)) {
          return SessionStore.fileBacked(s.path, s.displayName);
        }
      }
      throw new Error(`no session matching id: ${opts.sessionId}`);
    }
    const recent = await this.mostRecent(directory);
    if (recent) return SessionStore.fileBacked(recent.path, recent.displayName);
    return this.new(directory);
  }

  async mostRecent(directory: string): Promise<SessionSummary | null> {
    const sessions = await this.listSessions(directory);
    return sessions.reduce<SessionSummary | null>(
      (acc, s) => (acc === null || s.mtime > acc.mtime ? s : acc),
      null,
    );
  }

  async listSessions(directory: string): Promise<SessionSummary[]> {
    const d = join(this.base, this.slugFor(directory));
    let names: string[];
    try {
      names = await readdir(d);
    } catch {
      return [];
    }
    const out: SessionSummary[] = [];
    for (const name of names.filter((n) => n.endsWith(".jsonl")).sort()) {
      const p = join(d, name);
      try {
        const text = await readFile(p, "utf8");
        const nonempty = text.split(/\r?\n/).filter((l) => l.trim());
        const st = await stat(p);
        out.push({
          sessionId: basename(name, ".jsonl"),
          path: p,
          displayName: firstUserText(nonempty),
          mtime: st.mtimeMs / 1000,
          nEntries: nonempty.length,
          created: firstEntryTs(nonempty) ?? st.mtimeMs / 1000,
        });
      } catch {
        // skip unreadable
      }
    }
    return out.sort((a, b) => b.mtime - a.mtime); // most-recently-used first
  }
}

function firstEntryTs(lines: string[]): number | null {
  if (!lines.length) return null;
  try {
    const obj = JSON.parse(lines[0]!) as { ts?: unknown };
    return typeof obj.ts === "number" ? obj.ts : null;
  } catch {
    return null;
  }
}

function firstUserText(lines: string[]): string | null {
  for (const line of lines) {
    try {
      const obj = JSON.parse(line) as { type?: unknown; payload?: { text?: unknown } };
      if (obj.type !== "user") continue;
      const text = obj.payload?.text;
      if (typeof text !== "string") continue;
      const firstLine = text.split(/\r?\n/).find((l) => l.trim());
      if (!firstLine) continue;
      const name = firstLine.trim().replace(/\s+/g, " ").slice(0, 48);
      if (name) return name;
    } catch {
      // skip unparseable line
    }
  }
  return null;
}
